pkg/services/sqlstore/datasource.go

```diff
@@ -120,42 +120,40 @@ func (ss *SQLStore) DeleteDataSource(uid string, id int64, name string, orgID in
 	return cmd.DeletedDatasourcesCount, nil
 }
 
-// DeleteDataSource removes a datasource by org_id as well as either uid (preferred), id, or name
-// and is added to the bus.
+// DeleteDataSource removes a datasource by org_id as well as uid
 func DeleteDataSource(cmd *models.DeleteDataSourceCommand) error {
-	params := make([]interface{}, 0)
+	return inTransaction(func(sess *DBSession) error {
+		query := &models.GetDataSourceQuery{
+			Id:    cmd.ID,
+			Uid:   cmd.UID,
+			Name:  cmd.Name,
+			OrgId: cmd.OrgID,
+		}
 
-	makeQuery := func(sql string, p ...interface{}) {
-		params = append(params, sql)
-		params = append(params, p...)
-	}
+		if err := GetDataSource(query); err != nil {
+			cmd.DeletedDatasourcesCount = 0
+			return nil
```

Comment:
should this return err?

Reply:
good question. it's something i forgot to bring up.

it seemed like the existing behavior was to return 0 results without an error if the query didn't delete anything based on this test: https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/datasource_test.go#L224-L235

Reply:
Interesting, I see. It might be worthwhile to check `errors.Is(models.ErrDataSourceIdentifierNotSet)` and return the err in that case otherwise return nil.

Without that we're missing this case from the original:
```
	case cmd.OrgID == 0:
		return models.ErrDataSourceIdentifierNotSet
```

It might be cleaner to check that explicitly

Reply:
👍 good catch. i added that missing check in 0ae8cce with test coverage